recommendation/microsoft.owin.security.json

```diff
@@ -0,0 +1,907 @@
+                "NotContains": ""
+              }
+            }
+          ]
+        }
+      ]
+    },
+    {
+      "Type": "Method",
+      "Name": "Microsoft.Owin.Security.IAuthenticationManager.SignOut(params string[])",
+      "Value": "Microsoft.Owin.Security.IAuthenticationManager.SignOut(params string[])",
+      "KeyType": "",
+      "ContainingType": "IAuthenticationManager",
+      "RecommendedActions": [
+        {
+          "Source": "Amazon",
+          "Preferred": "Yes",
+          "TargetFrameworks": [
+            {
+              "Name": "netcoreapp3.1",
+              "TargetCPU": [ "x86", "x64", "ARM32", "ARM64" ]
+            },
+            {
+              "Name": "net5.0",
+              "TargetCPU": [ "x86", "x64", "ARM32", "ARM64" ]
+            }
+          ],
+          "Description": "For Microsoft.Owin.Security.IAuthenticationManager.SignOut(params string[]) add directive Microsoft.AspNetCore.Authentication, add a comment to explain parameters, add an await operator and replace the method SignOut with SignOutAsync.",
+          "Actions": [
+            {
+              "Name": "AddDirective",
+              "Type": "Using",
+              "Value": "Microsoft.AspNetCore.Authentication",
+              "Description": "Add Microsoft.AspNetCore.Authentication namespace",
+              "ActionValidation": {
+                "Contains": "using Microsoft.AspNetCore.Authentication;",
+                "NotContains": ""
+              }
+            },
+            {
+              "Name": "ReplaceMethodAndParameters",
+              "Type": "Method",
+              "Value": {
+                "oldMethod": "SignOut",
+                "newMethod": "SignOutAsync",
+                "newParameters": "()"
+              },
+              "Description": "Replace method name from SignOut to SignOutAsync and remove all parameters.",
+              "ActionValidation": {
+                "Contains": "SignOutAsync",
+                "NotContains": ""
+              }
+            },
+            {
+              "Name": "AddComment",
+              "Type": "Method",
+              "Value": "You can only pass in a single scheme as a parameter and you can also include an AuthenticationProperties object as well.",
+              "Description": "Add a comment to explain how to use the parameters for the SignOutAsync api.",
+              "ActionValidation": {
+                "Contains": "You can only pass in a single scheme as a parameter and you can also include an AuthenticationProperties object as well.",
+                "NotContains": "",
+                "CheckComments": "True"
+              }
+            },
+            {
+              "Name": "AddAwaitOperator",
+              "Type": "Expression",
+              "Value": "",
+              "Description": "Add the await operator to SignOutAsync.",
+              "ActionValidation": {
+                "Contains": "await",
+                "NotContains": ""
```

Comment:
Since `await` is being added to `SignOut`/`SignOutAsync`, would it make the validation more robust if `Contains` were: `"await SignOut"`?

Reply:
Since the rules for validation are per action based and the add await operator depends on another action I didn't want to assume the other action was also successful and wanted to test them independently. Although granted testing for the presence of an await qualifier is not super useful in and of itself. I know we also conduct all validation checks after all actions have been applied but if we ever change that around and we check after each action making this change would also force 2 actions' validation together. What do you think?

Reply:
I was thinking that in either case where both actions take place or only Action 2 takes place, the validation of `"Contains": "await SignOut"` would still be valid. This validates the `AddAwaitOperator` action by itself while still remaining valid whether or not the other action takes place. This would also make it more difficult to give a false positive on the validation (unless I'm missing something).

Reply:
This would also apply to `SignIn/SignInAsync` and `Challenge/ChallengeAsync` which I missed in my initial review

Reply:
Thanks Jon, agreed that validation should be tightly coupled with rule as a whole rather than action and any validation logic change would invalidate most validation logic anyway. Made the change.